### PR TITLE
Set total_rows_approx for trivial queries with LIMIT from system.zeros and generateRandom

### LINKED ISSUE #58183: If the storage does not provide `total_rows_approx` for the progress bar, but it is a trivial query with LIMIT, provide `total_rows_approx` from the LIMIT.
**Use case**

This query should show progress bar:
```
INSERT INTO test SELECT * FROM generateRandom() LIMIT 100000000000;
```

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The progress bar will work for trivial queries with LIMIT from `system.zeros`, `system.zeros_mt` (it already works for `system.numbers` and `system.numbers_mt`), and the `generateRandom` table function. As a bonus, if the total number of records is greater than the `max_rows_to_read` limit, it will throw an exception earlier. This closes #58183.

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Set desired options before CI starts or re-push after updates

#### Run only:
- [ ]  Integration tests
- [ ]  Integration tests (arm64)
- [ ]  Stateless tests (release)
- [ ]  Stateless tests (asan)
- [ ]  Stateful tests (release)
- [ ]  Stateful tests (asan)
- [ ]  No sanitizers
- [ ]  Tests with analyzer
- [ ]  Fast tests
- [ ]  Only package_debug build
- [ ]  Add your CI variant description here

#### CI options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 86039802e3327c0a527135233f3589d6bcea2348 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61823/86039802e3327c0a527135233f3589d6bcea2348/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 1]
```diff
@@ -2499,10 +2499,13 @@ void InterpreterSelectQuery::executeFetchColumns(QueryProcessingStage::Enum proc
             max_block_size = std::max<UInt64>(1, max_block_limited);
             max_threads_execute_query = max_streams = 1;
         }
+
         if (local_limits.local_limits.size_limits.max_rows != 0)
         {
             if (max_block_limited < local_limits.local_limits.size_limits.max_rows)
                 query_info.limit = max_block_limited;
+            else if (local_limits.local_limits.size_limits.max_rows < std::numeric_limits<UInt64>::max()) /// Ask to read just enough rows to make the max_rows limit effective (so it has a chance to be triggered).
```
Comment (vdimir) [inline src/Interpreters/InterpreterSelectQuery.cpp:2507]: In new analyzer the same logic is here
https://github.com/ClickHouse/ClickHouse/blob/bf2edb972e897a9993e2a047c9666514c595876c/src/Planner/PlannerJoinTree.cpp#L689-L694

Setting `table_expression_query_info.limit` there fixes `03024_total_rows_approx_is_set_for_system_zeros_and_generate_random`

**Comment by novikd** [comment]
Let's not modify `InterpreterSelectQuery.cpp`. We should not add changes to the old infrastructure after we enabled analyzer by default.

**Comment by alexey-milovidov** [comment]
It was not my intent to add it to the old analyzer (only to the new), but as soon as I accidentally did it, let's keep it.

**Comment by clickgapai** [comment]
Hi @alexey-milovidov — while reviewing this PR I found the following:

- **Bug (P3):** `trivial_limit = 1 + max_rows` triggers `TOO_MANY_ROWS` for boundary case `LIMIT N` with `max_rows_to_read = N` — https://github.com/ClickHouse/ClickHouse/issues/103956

Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**